In [1]:
import pandas as pd

In [2]:
data = pd.read_csv('train_values.csv')
target = pd.read_csv('train_labels.csv')

In [3]:
df = pd.merge(data, target, how='inner')

In [4]:
data.head()

patient_id  slope_of_peak_exercise_st_segment               thal  \
0     0z64un                                  1             normal   
1     ryoo3j                                  2             normal   
2     yt1s1x                                  1             normal   
3     l2xjde                                  1  reversible_defect   
4     oyt4ek                                  3  reversible_defect   

   resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                     128                2                  0   
1                     110                3                  0   
2                     125                4                  3   
3                     152                4                  0   
4                     178                1                  0   

   fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                     0                    2   
1                                     0                    0   
2                                     0                    2   
3                                     0                    0   
4                                     0                    2   

   serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                          308                       0.0    1   45   
1                          214                       1.6    0   54   
2                          304                       0.0    1   77   
3                          223                       0.0    1   40   
4                          270                       4.2    1   59   

   max_heart_rate_achieved  exercise_induced_angina  
0                      170                        0  
1                      158                        0  
2                      162                        1  
3                      181                        0  
4                      145                        0

In [5]:
target.head()

patient_id  heart_disease_present
0     0z64un                      0
1     ryoo3j                      0
2     yt1s1x                      1
3     l2xjde                      1
4     oyt4ek                      0

# EDA

In [6]:
target['heart_disease_present'].value_counts()

0    100
1     80
Name: heart_disease_present, dtype: int64

In [7]:
data.shape

(180, 14)

In [8]:
target.shape

(180, 2)

In [9]:
df.shape

(180, 15)

Checking for NULL values

In [10]:
data.isnull().sum(axis=0)

patient_id                              0
slope_of_peak_exercise_st_segment       0
thal                                    0
resting_blood_pressure                  0
chest_pain_type                         0
num_major_vessels                       0
fasting_blood_sugar_gt_120_mg_per_dl    0
resting_ekg_results                     0
serum_cholesterol_mg_per_dl             0
oldpeak_eq_st_depression                0
sex                                     0
age                                     0
max_heart_rate_achieved                 0
exercise_induced_angina                 0
dtype: int64

In [11]:
data.describe()

slope_of_peak_exercise_st_segment  resting_blood_pressure  \
count                         180.000000              180.000000   
mean                            1.550000              131.311111   
std                             0.618838               17.010443   
min                             1.000000               94.000000   
25%                             1.000000              120.000000   
50%                             1.000000              130.000000   
75%                             2.000000              140.000000   
max                             3.000000              180.000000   

       chest_pain_type  num_major_vessels  \
count       180.000000         180.000000   
mean          3.155556           0.694444   
std           0.938454           0.969347   
min           1.000000           0.000000   
25%           3.000000           0.000000   
50%           3.000000           0.000000   
75%           4.000000           1.000000   
max           4.000000           3.000000   

       fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
count                            180.000000           180.000000   
mean                               0.161111             1.050000   
std                                0.368659             0.998742   
min                                0.000000             0.000000   
25%                                0.000000             0.000000   
50%                                0.000000             2.000000   
75%                                0.000000             2.000000   
max                                1.000000             2.000000   

       serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression         sex  \
count                   180.000000                180.000000  180.000000   
mean                    249.211111                  1.010000    0.688889   
std                      52.717969                  1.121357    0.464239   
min                     126.000000                  0.000000    0.000000   
25%                     213.750000                  0.000000    0.000000   
50%                     245.500000                  0.800000    1.000000   
75%                     281.250000                  1.600000    1.000000   
max                     564.000000                  6.200000    1.000000   

              age  max_heart_rate_achieved  exercise_induced_angina  
count  180.000000               180.000000               180.000000  
mean    54.811111               149.483333                 0.316667  
std      9.334737                22.063513                 0.466474  
min     29.000000                96.000000                 0.000000  
25%     48.000000               132.000000                 0.000000  
50%     55.000000               152.000000                 0.000000  
75%     62.000000               166.250000                 1.000000  
max     77.000000               202.000000                 1.000000

In [12]:
df = df.drop('patient_id', axis=1)

In [13]:
df.head()

slope_of_peak_exercise_st_segment               thal  \
0                                  1             normal   
1                                  2             normal   
2                                  1             normal   
3                                  1  reversible_defect   
4                                  3  reversible_defect   

   resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                     128                2                  0   
1                     110                3                  0   
2                     125                4                  3   
3                     152                4                  0   
4                     178                1                  0   

   fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                     0                    2   
1                                     0                    0   
2                                     0                    2   
3                                     0                    0   
4                                     0                    2   

   serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                          308                       0.0    1   45   
1                          214                       1.6    0   54   
2                          304                       0.0    1   77   
3                          223                       0.0    1   40   
4                          270                       4.2    1   59   

   max_heart_rate_achieved  exercise_induced_angina  heart_disease_present  
0                      170                        0                      0  
1                      158                        0                      0  
2                      162                        1                      1  
3                      181                        0                      1  
4                      145                        0                      0

In [14]:
data.dtypes

patient_id                               object
slope_of_peak_exercise_st_segment         int64
thal                                     object
resting_blood_pressure                    int64
chest_pain_type                           int64
num_major_vessels                         int64
fasting_blood_sugar_gt_120_mg_per_dl      int64
resting_ekg_results                       int64
serum_cholesterol_mg_per_dl               int64
oldpeak_eq_st_depression                float64
sex                                       int64
age                                       int64
max_heart_rate_achieved                   int64
exercise_induced_angina                   int64
dtype: object

Since 'thal' is an object datatype, we need to one hot encode it

In [15]:
df = pd.get_dummies(df, drop_first=True)

In [16]:
df.head()

slope_of_peak_exercise_st_segment  resting_blood_pressure  chest_pain_type  \
0                                  1                     128                2   
1                                  2                     110                3   
2                                  1                     125                4   
3                                  1                     152                4   
4                                  3                     178                1   

   num_major_vessels  fasting_blood_sugar_gt_120_mg_per_dl  \
0                  0                                     0   
1                  0                                     0   
2                  3                                     0   
3                  0                                     0   
4                  0                                     0   

   resting_ekg_results  serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  \
0                    2                          308                       0.0   
1                    0                          214                       1.6   
2                    2                          304                       0.0   
3                    0                          223                       0.0   
4                    2                          270                       4.2   

   sex  age  max_heart_rate_achieved  exercise_induced_angina  \
0    1   45                      170                        0   
1    0   54                      158                        0   
2    1   77                      162                        1   
3    1   40                      181                        0   
4    1   59                      145                        0   

   heart_disease_present  thal_normal  thal_reversible_defect  
0                      0            1                       0  
1                      0            1                       0  
2                      1            1                       0  
3                      1            0                       1  
4                      0            0                       1

In [17]:
df_Y = df['heart_disease_present']
df_X = df.drop('heart_disease_present', axis=1)

In [18]:
df_Y.head()

0    0
1    0
2    1
3    1
4    0
Name: heart_disease_present, dtype: int64

In [19]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier

In [20]:
from sklearn.cross_validation import cross_val_score

C:\Users\absol\Anaconda3.4\lib\site-packages\sklearn\cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [21]:
import numpy as np

In [22]:
cross_val_score(LogisticRegression(), df_X, df_Y, cv=5, scoring='neg_log_loss')

array([-0.53283297, -0.32564027, -0.21982994, -0.59801933, -0.40967314])

In [23]:
np.abs(cross_val_score(LogisticRegression(), df_X, df_Y, cv=10, scoring='neg_log_loss').mean())

0.42349757512849512

In [24]:
np.abs(cross_val_score(SVC(probability=True), df_X, df_Y, cv=10, scoring='neg_log_loss').mean())

0.69464540683323639

In [25]:
cross_val_score(RandomForestClassifier(n_estimators=1000), df_X, df_Y, cv=5, scoring='neg_log_loss').mean()

-0.43024067187603104

In [26]:
import xgboost

In [27]:
cross_val_score(xgboost.XGBClassifier(n_estimators=100, ), df_X, df_Y, cv=5, scoring='neg_log_loss').mean()

-0.45701502819752526

In [28]:
xgb = xgboost.XGBClassifier()

In [29]:
xgb.fit(df_X, df_Y)

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
       colsample_bytree=1, gamma=0, learning_rate=0.1, max_delta_step=0,
       max_depth=3, min_child_weight=1, missing=None, n_estimators=100,
       n_jobs=1, nthread=None, objective='binary:logistic', random_state=0,
       reg_alpha=0, reg_lambda=1, scale_pos_weight=1, seed=None,
       silent=True, subsample=1)

In [30]:
clr = LogisticRegression()

In [31]:
clr.fit(df_X, df_Y)

LogisticRegression(C=1.0, class_weight=None, dual=False, fit_intercept=True,
          intercept_scaling=1, max_iter=100, multi_class='ovr', n_jobs=1,
          penalty='l2', random_state=None, solver='liblinear', tol=0.0001,
          verbose=0, warm_start=False)

In [32]:
test_data = pd.read_csv('test_values.csv')

In [33]:
test_data.head()

patient_id  slope_of_peak_exercise_st_segment               thal  \
0     olalu7                                  2  reversible_defect   
1     z9n6mx                                  1             normal   
2     5k4413                                  2  reversible_defect   
3     mrg7q5                                  1             normal   
4     uki4do                                  2             normal   

   resting_blood_pressure  chest_pain_type  num_major_vessels  \
0                     170                1                  0   
1                     138                4                  0   
2                     120                4                  0   
3                     102                3                  1   
4                     138                4                  1   

   fasting_blood_sugar_gt_120_mg_per_dl  resting_ekg_results  \
0                                     0                    2   
1                                     0                    0   
2                                     0                    2   
3                                     0                    0   
4                                     0                    2   

   serum_cholesterol_mg_per_dl  oldpeak_eq_st_depression  sex  age  \
0                          288                       0.2    1   59   
1                          183                       1.4    0   35   
2                          177                       2.5    1   43   
3                          318                       0.0    0   60   
4                          166                       3.6    1   61   

   max_heart_rate_achieved  exercise_induced_angina  
0                      159                        0  
1                      182                        0  
2                      120                        1  
3                      160                        0  
4                      125                        1

In [34]:
test_X = test_data.drop('patient_id', axis=1)

In [35]:
test_X = pd.get_dummies(test_X, drop_first=True)

In [36]:
y_pred = xgb.predict_proba(test_X)

In [37]:
submission = pd.DataFrame()

In [38]:
submission['patient_id'] = test_data['patient_id']

In [39]:
submission['heart_disease_present'] = y_pred[:,1]

In [40]:
submission.to_csv('baseline_submission_2.csv', index=False)

We used XGBoost and Logistic Regression for our baseline models and got into the top 15% in the competition